src/games/strategy/triplea/ui/mapdata/RouteOptimizer.java

```diff
@@ -0,0 +1,185 @@
+package games.strategy.triplea.ui.mapdata;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import java.awt.Point;
+import java.awt.geom.Line2D;
+import java.awt.geom.Point2D;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import games.strategy.triplea.ui.MapPanel;
+
+public class RouteOptimizer {
+
+  public final boolean isInfiniteY;
+  public final boolean isInfiniteX;
+
+  private final MapPanel mapPanel;
+
+  private Point endPoint;
+  private int mapWidth;
+  private int mapHeight;
+
+  public RouteOptimizer(MapData mapData, MapPanel mapPanel) {
+    checkNotNull(mapData);
+    this.mapPanel = checkNotNull(mapPanel);
+    isInfiniteY = mapData.scrollWrapY();
+    isInfiniteX = mapData.scrollWrapX();
+  }
+
+  public Point[] getTranslatedRoute(Point... route) {
+    if (route == null) {
+      return route;
+    }
+    if (!isInfiniteX && !isInfiniteY || route.length == 0) {
```

Comment:
Please put parens to do explicit grouping of the boolean conditions, ie: Does 'a & b || c' mean "(a & b) || c", or is it "a & (b || c)"